### PR TITLE
ci: upload gzipped pkg binaries

### PR DESCRIPTION
**Pull request by johnpc**
#### Description of changes

Since shipping Amplify CLI 8.0.0, we're vending platform specific binaries downloaded as a post install script of `@aws-amplify/cli` npm package. 

This commit adds compression to the binaries to reduce network overhead of that download, speeding up install times, reducing variability caused by different network conditions, and overall limiting impact on bandwidth for CLI users. A ~72% improvement!

Linux: 696 MB to 197 MB
MacOS: 701 MB to 198 MB
Windows: 690 MB to 195 MB

#### Issue #, if available

n/a

#### Description of how you validated changes

* cloud-e2e: https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli?branch=run-e2e%2Fxss%2Fbinary-compression&filter=all
* tagged release: https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli?branch=tagged-release-without-e2e-tests%2Fbinary-compression-2&filter=all

#### Checklist

- [x] PR description included
- [x] `yarn test` passes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-cli-npm/binary.ts [hunk 1]
```diff
@@ -101,37 +101,37 @@ export class Binary {
...
-      mkdirSync(this.installDirectory, { recursive: true });
+    if (!fs.existsSync(this.installDirectory)) {
+      fs.mkdirSync(this.installDirectory, { recursive: true });
     }
 
     let amplifyExecutableName = 'amplify';
-    if (getPlatformBinaryName().endsWith('.exe')) {
+    if (os.type() === 'Windows_NT') {
       amplifyExecutableName = 'amplify.exe';
     }
```
Comment (jhockett) [inline packages/amplify-cli-npm/binary.ts:113]: nit: I'd sacrifice readability to make `amplifyExecutableName` a `const`, but either way is fine.

```suggestion
    const amplifyExecutableName = os.type() === 'Windows_NT' ? 'amplify.exe' : 'amplify';
```

File: packages/amplify-cli-npm/binary.ts [hunk 2]
```diff
@@ -152,4 +152,35 @@ export class Binary {
...
+
+  /**
+   * Extracts a .tar file
+   *
+   * @returns tar.Extract
+   */
+  private extract(): tar.Extract {
+    const extract = tar.extract();
+    const chunks: Uint8Array[] = [];
+    extract.on('entry', (header, s, next) => {
```
Comment (jhockett) [inline packages/amplify-cli-npm/binary.ts:164]: Please use a more descriptive name than `s`.

**Review by jhockett — APPROVED** [review]

File: packages/amplify-cli-npm/binary.ts [hunk 3]
```diff
@@ -101,37 +101,37 @@ export class Binary {
...
     }
 
-    mkdirSync(this.installDirectory, { recursive: true });
-    console.log(`Downloading release from ${getBinaryUrl()}`);
+    fs.mkdirSync(this.installDirectory, { recursive: true });
+    console.log(`Downloading release from ${getCompressedBinaryUrl()}`);
     try {
-      const res = await axios({ url: getBinaryUrl(), responseType: 'stream' });
+      const res = await axios({ url: getCompressedBinaryUrl(), responseType: 'stream' });
       await pipeline(
```
[the remark this answers is not in the document]
Comment (johnpc) [reply packages/amplify-cli-npm/binary.ts:126]: worth noting that every time we add a dependency like this, it increases our supply chain attack surface.

[the remark this answers is not in the document]
Comment (johnpc) [reply packages/amplify-cli-npm/binary.ts:126]: Also the behavior here is different. Using 

`const downloadPromise = promisify(pipeline)(response.body, gunzip(), tar.extract(binDir));` extracts the entire directory - not just the file. So you get `~/.amplify/bin/amplify/amplify.exe` for example.

Using the code added in this PR, you have one fewer dependency and it results in the proper structure `~/.amplify/bin/amplify.exe` for example.

**Pull request merged**